doc/life-of-a-request.md

```diff
@@ -0,0 +1,164 @@
+
+If the connection begins with a HTTP/2 Prior-knowledge preface, the connection is handled
+as HTTP/2.
+
+
+## TCP Forwarding
+
+If, after protocol detection (and potentially TLS decryption), the connection's protocol
+is not supported by the proxy, the connection is simply forwarded to the original
+destination.
+
+Policy may require that the forwarded connections are TLS'd by the proxy (i.e. for mutual
+authentication).
+
+
+## HTTP Proxying
+
+### HTTP/1.0
+
+If the request is an HTTP/1.0 request that does not include a `Host` header, the request
+is routed to the connection's original destination address. The request is sent on a
+dedicated one-off connection (in keeping with the HTTP/1.0 specification).
+
+If an HTTP/1.0 request is received and _does_ include a `Host` header, the request is
+routed as if it were HTTP/1.1, though the proxy must be sure to honor proper HTTP/1.0
+semantics to the client peer.
+
+### HTTP/1.1
+
+When the proxy receives an HTTP/1.1 request, it is checked to see if the `Connection`
+header includes `Upgrade` in its list of values. If the `Upgrade` header exists, and its
+value is `h2c`, the connection is upgraded to HTTP/2 before continuing. If it has any
+other value, then the request should fail with a 500[1](#http-1).
+
+The proxy honors hop-by-hop headers (i.e. `Connection`, `Authorization`) and it removes
+them from requests before they are routed.
+
+The proxy honors `CONNECT` requests as any other. Once the request is routed to an
+appropriate destination, it is dispatched, and the accepted connection is treated as part
+of the request.
+
+All HTTP/1.1 requests must have a valid `Host` value.
+
+### HTTP/2
+
+The proxy honors `CONNECT` requests as any other.
+
+The proxy honors hop-by-hop headers (i.e. `Connection`, `Authorization`) and it removes
+them from requests before they are routed.
+
+All HTTP/2 requests must have a valid `:authority` value.
+
+### Routing
+
+HTTP/2 requests are routed by their `:authority`, and HTTP/1 requests by their `Host`.
+We'll refer to all of these as an _Authority_.
+
+#### Service Discovery
+
+In order to perform request-level load balancing, the proxy needs to resolve the Authority
+to a set of TCP addresses.
+
+The proxy issues a request to the Conduit control plane with the _Authority_, as specified
+in the request, as well as a set of _destination search paths_, specified to proxy at
+start-time. The control plane attempts to resolve the authority (using its search paths)
+through its known service discovery backends.
+
+For example, suppose we attempt to route a request with an authority of _rabbits_. If the
+proxy's running in a kubernetes cluster in the _farm_ namespace, the proxy should be
+configured with a destination search path like:
+```
+farm.svc.cluster.local.
+svc.cluster.local.
+```
+
+The control plane attempts to discover endpoints via the names:
+1. `rabbits.`
+2. `rabbits.farm.svc.cluster.local.`
+3. `rabbits.svc.cluster.local.`
```

Comment:
I don't think this is right. Instead I propose the following: The proxy gets a configuration parameter, e.g. `DESTINATION_SERVICE_DNS_SUFFIX=.svc.cluster.local.` that indicates which DNS suffix should be handled by the controller. Then the proxy will look up the search domains in /etc/resolv.conf and go through the list trying to complete the input name for each suffix. If the name completes to a string ending in `$DESTINATION_SERVICE_DNS_SUFFIX` then the destination service will be asked about the full name. Otherwise, we'll fall back to the fallback mechanism (DNS currently).

This will handle every single DNS configuration in Kubernetes correctly.

Note that initially we might just assume that the search list is `.$current_namespace.svc.cluster.local. .svc.cluster.local, .cluster.local` so that we don't have to read it out of `/etc/resolv.conf`. OTOH, we already have the search list available to us since the `domain` crate makes it available, so eventually we'll be able to implement the correct full semantics, even if we take a shortcut to start.

Reply:
@briansmith Keeping in mind that I'm trying to describe some end state of the proxy API where it may be usable with a fuller or different controller implementation (slightly divorced from the next steps in execution, but hopefully useful towards API design):

> The proxy gets a configuration parameter, e.g. DESTINATION_SERVICE_DNS_SUFFIX=.svc.cluster.local. that indicates which DNS suffix should be handled by the controller. Then the proxy will look up the search domains in /etc/resolv.conf and go through the list trying to complete the input name for each suffix. If the name completes to a string ending in $DESTINATION_SERVICE_DNS_SUFFIX then the destination service will be asked about the full name. Otherwise, we'll fall back to the fallback mechanism (DNS currently).

This approach sounds good, though I think that this should probably be a list of non-overlapping FQDN suffixes so that we don't have API-level assumptions about the controller supporting only one discovery scheme.

> Note that initially we might just assume that the search list is .$current_namespace.svc.cluster.local. .svc.cluster.local, .cluster.local so that we don't have to read it out of /etc/resolv.conf. OTOH, we already have the search list available to us since the domain crate makes it available, so eventually we'll be able to implement the correct full semantics, even if we take a shortcut to start.

This raises an interesting API question: does the conduit proxy API allow service discovery through schemes that aren't exposed via DNS? I'm inclined to think that this shouldn't be a hard requirement (as you point above -- we're not guaranteed that the client did DNS resolution on the name previously). If that's the case, we'll probably end up with some concept of a conduit-specific search path that precedes /etc/resolv.conf so the discovery api  could potentially bind through additional service discovery backends (i.e. mapped by suffix).

Reply:
> This approach sounds good, though I think that this should probably be a list of non-overlapping FQDN suffixes so that we don't have API-level assumptions about the controller supporting only one discovery scheme.

Agreed.

> This raises an interesting API question: does the conduit proxy API allow service discovery through schemes that aren't exposed via DNS? I'm inclined to think that this shouldn't be a hard requirement (as you point above -- we're not guaranteed that the client did DNS resolution on the name previously). If that's the case, we'll probably end up with some concept of a conduit-specific search path that precedes /etc/resolv.conf so the discovery api could potentially bind through additional service discovery backends (i.e. mapped by suffix).

I agree. I think we'll be able to easily cope with having some source other than, or in addition to, `/etc/resolv.conf` as the source for the search path, without having to design it in detail now.